usaspending_api/common/etl/spark.py

```diff
@@ -560,3 +564,177 @@ def create_ref_temp_views(spark: SparkSession, create_broker_views: bool = False
+    """
+    if not logger:
+        logger = get_jvm_logger(spark)
+    # Delete output data dir if it already exists
+    parts_dir_path = spark.sparkContext._jvm.org.apache.hadoop.fs.Path(parts_dir)
+    fs = parts_dir_path.getFileSystem(spark.sparkContext._jsc.hadoopConfiguration())
+    if fs.exists(parts_dir_path):
+        fs.delete(parts_dir_path, True)
+    start = time.time()
+    logger.info(f"Writing source data DataFrame to csv part files for file {parts_dir}...")
+    df_record_count = df.count()
+    df.write.options(
+        # NOTE: this is a suggestion, to be used by Spark if partitions yield multiple files
+        maxRecordsPerFile=max_rows_per_merged_file,
+    ).csv(
+        path=parts_dir,
+        header=False,
+        nullValue=None,
+        escape='"',  # " is used to escape the 'quote' character setting (which defaults to "). Escaped quote = ""
+        ignoreLeadingWhiteSpace=False,  # must set for CSV write, as it defaults to true
+        ignoreTrailingWhiteSpace=False,  # must set for CSV write, as it defaults to true
+        timestampFormat=CONFIG.SPARK_CSV_TIMEZONE_FORMAT,
+        mode="overwrite" if overwrite else "errorifexists",
+    )
+    logger.info(f"{parts_dir} contains {df_record_count:,} rows of data")
+    logger.info(f"Wrote source data DataFrame to csv part files in {(time.time() - start):3f}s")
+    return df_record_count
+
+
+def hadoop_copy_merge(
+    spark: SparkSession,
+    parts_dir: str,
+    header: str,
+    max_rows_per_merged_file=EXCEL_ROW_LIMIT,
+    logger=None,
+    file_format=".csv",
+) -> List[str]:
+    """PySpark impl of Hadoop 2.x copyMerge() (deprecated in Hadoop 3.x)
+    Merges files from a provided input directory and then redivides them
+        into multiple files based on a provided maximum row count parameter.
+    Args:
+        spark: passed-in active SparkSession
+        parts_dir: Path to the dir that contains the input parts files. The parts dir name
+            determines the name of the merged files.
+        header: A comma-separated list of field names, to be placed as the first row of every final CSV file.
+            Individual part files must NOT therefore be created with their own header.
+        max_rows_per_merged_file: Final CSV data will be subdivided into numbered files so that there is not more than
+            this many rows in any file written. Only if the total data exceeds this value will multiple files be
+                created with a pattern of ``{merged_file}_N.{extension}`` with N starting at 1.
+        logger: The logger to use. If one note provided (e.g. to log to console or stdout) the underlying JVM-based
+            Logger will be extracted from the ``spark`` ``SparkSession`` and used as the logger.
+        file_format: The format of the part files and the format of the final merged file
+
+    Returns:
+        a list of file names where each element is a merged file that was generated during the copy merge.
```

Comment:
For consistency, I'd like to see the leading `a` capitalized

Reply:
I've addressed, but we need to include this in our style guide if this is something we expect people to follow.

Reply:
Here is an example of code that doesn't follow this style, it's existed for 3 years.
```
def _lookup_deleted_award_keys(
    client: Elasticsearch,
    lookup_key: str,
    value_list: list,
    config: dict,
    index: Optional[str] = None,
    lookup_chunk_size: int = 50000,
) -> list:
    """Derive a list of award keys given a target index, Lookup field, and lookup values

    This returns a list of all unique award keys, which are compiled from the ``ES_AWARDS_UNIQUE_KEY_FIELD`` field of
    any document in the given ``index`` that matches the query. The matching query is a terms query that will return
    the doc if its ``lookup_key`` field has any value provided in ``value_list``.

    Args:
        client (Elasticsearch): elasticsearch-dsl client for making calls to an ES cluster
        lookup_key (str): name of field in targeted elasticsearch index by which we are looking up docs. The field or
            sub-field provided MUST be of ``keyword`` type (or ``_id`` meta field)
        value_list (list): if lookup_key field has any of these values, the document will be returned from the lookup
        config (dict): collection of key-value pairs that encapsulates runtime arguments for this ES management task
        index (str): Optional name, alias, or pattern of index this query will target. Looks up via config if not
            provided
        lookup_chunk_size (int): the batch-size of terms value-array to be looked-up. Needs to be less
            than 65536 (max values for any terms query), and less than config["max_query_size"]

    Returns: list of values for the ES_AWARDS_UNIQUE_KEY_FIELD fields in the looked-up documents.
    """
```

Reply:
Here's another example
```
def extract_location(recipient_hash):
    """Extract the location data via the recipient hash

    Args:
        recipient_hash: uuid of the hash+duns to look up

    Returns:
        dict of location info
    """
```